src/cpp/writer/tswriter.cpp

```diff
@@ -13,123 +14,164 @@ TsWriterCPP::TsWriterCPP(
     }
-    case (32): {
-        auto data_array = tensorstore::Array(py_image.mutable_unchecked<std::int16_t, 1>().data(0), _image_shape, tensorstore::c_order);
 
-        auto write_result = tensorstore::Write(tensorstore::UnownedToShared(data_array), _source).result();
-        if (!write_result.ok()) {
-            std::cerr << "Error writing image: " << write_result.status() << std::endl;
-        }
-        break;   
+    if (_c_index.has_value() && channels.has_value()) {
+        output_transform = (std::move(output_transform) | tensorstore::Dims(_c_index.value()).ClosedInterval(channels.value().Start(), channels.value().Stop())).value();
     } 
-    case (64): {
-        auto data_array = tensorstore::Array(py_image.mutable_unchecked<std::int32_t, 1>().data(0), _image_shape, tensorstore::c_order);
 
-        auto write_result = tensorstore::Write(tensorstore::UnownedToShared(data_array), _source).result();
-        if (!write_result.ok()) {
-            std::cerr << "Error writing image: " << write_result.status() << std::endl;
-        }
-        break;
-    }
-    case (128): {
-        auto data_array = tensorstore::Array(py_image.mutable_unchecked<std::int64_t, 1>().data(0), _image_shape, tensorstore::c_order);
+    if (_t_index.has_value() && tsteps.has_value()) {
+        output_transform = (std::move(output_transform) | tensorstore::Dims(_t_index.value()).ClosedInterval(tsteps.value().Start(), tsteps.value().Stop())).value();
+    } 
 
-        // Write data array to TensorStore
-        auto write_result = tensorstore::Write(tensorstore::UnownedToShared(data_array), _source).result();
-        if (!write_result.ok()) {
-            std::cerr << "Error writing image: " << write_result.status() << std::endl;
-        }
-        break;
-    }
-    case (256): {
-        auto data_array = tensorstore::Array(py_image.mutable_unchecked<float, 1>().data(0), _image_shape, tensorstore::c_order);
+    output_transform = (std::move(output_transform) | tensorstore::Dims(_y_index).ClosedInterval(rows.Start(), rows.Stop()) |
+                                                      tensorstore::Dims(_x_index).ClosedInterval(cols.Start(), cols.Stop())).value(); 
 
-        auto write_result = tensorstore::Write(tensorstore::UnownedToShared(data_array), _source).result();
-        if (!write_result.ok()) {
-            std::cerr << "Error writing image: " << write_result.status() << std::endl;
-        }
-        break;
-    }
-    case (512): {
-        auto data_array = tensorstore::Array(py_image.mutable_unchecked<double, 1>().data(0), _image_shape, tensorstore::c_order);
+    // use switch instead of template to avoid creating functions for each datatype
+    switch(_dtype_code)
+    {
+        case (1): {
+            auto data_array = tensorstore::Array(py_image.mutable_unchecked<std::uint8_t, 1>().data(0), _image_shape, tensorstore::c_order);
```

Comment:
I don't think `_image_shape` is correct here. What if the data we are writing is only a part of the image? IMO, it should be deduced from `rows`, `cols`...etc.